hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1612 @@
+    pub fn set_urad_int(self) {
+        self.token.ier().write(|w| unsafe { w.rsu().bits(1) });
+    }
+
+    /// Disable Undefined Register Access Detection interrupt enable
+    #[inline]
+    pub fn disable_urad_int(self) {
+        self.token.idr().write(|w| unsafe { w.rsu().bits(1) });
+    }
+
+    /// Get Undefined Register Access Detection interrupt mask
+    #[inline]
+    pub fn get_urad_int_mask(&self) -> bool {
+        self.token.imr().read().urad().bits()
+    }
+
+    /// Get Undefined Register Access Detection interrupt status
+    #[inline]
+    pub fn get_urad_int(&self) -> bool {
+        self.token.isr().read().urad().bits()
+    }
+
+    /// Get Undefined Register Access Trace
+    ///
+    /// This field is only reset by `swrst`
+    #[inline]
+    pub fn get_urat(&self) -> Variant<u8, URAT_A> {
+        self.token.uasr().read().urat().variant()
+    }
+
+    /// When reading the interrupt (ISR) register, it is cleared
+    ///
+    /// This might result in unexpected results for example
+    /// when sequentially trying to determine which interrupt triggered.
+    ///
+    /// This is an alternative, return all the data from the register
+    /// and parse later with the designated `get_[name]_int` functions.
+    #[inline]
+    pub fn get_interrupt_status(&mut self) -> IcmInterrupt {
+        let interrupt_vector = self.token.isr().read().bits();
+        IcmInterrupt { interrupt_vector }
+    }
+    /// Trigger recalculation of memory monitor region specified
+    /// by the bitmask:
+    /// 0b0001 = region0
+    /// 0b0010 = region1
+    /// 0b0100 = region2
+    /// 0b1000 = region3
+    /// ...
+    /// 0b1111 = all regions
+    #[inline]
+    pub fn trigger_rehash(&self, bitmask: u8) {
+        self.token
+            .ctrl()
+            .write(|w| unsafe { w.rehash().bits(bitmask) });
+    }
+}
+
+/// Trait providing numerical identifier and
+/// offset for each ICM Region
+///
+/// ICM supports 4 memory regions
+pub trait IcmRegionNum: Sealed {
+    /// Numerical ID of the memory region
+    const NUM: usize;
+    /// Memory offset
+    const OFFSET: u32;
+}
+
+/// ICM Region 0
+pub enum Region0 {}
+impl Sealed for Region0 {}
+impl IcmRegionNum for Region0 {
+    const NUM: usize = 0;
+    const OFFSET: u32 = 0;
+}
+
+/// ICM Region 1
+pub enum Region1 {}
+impl Sealed for Region1 {}
+impl IcmRegionNum for Region1 {
+    const NUM: usize = 1;
+    const OFFSET: u32 = 0x10;
+}
+
+/// ICM Region 2
+pub enum Region2 {}
+impl Sealed for Region2 {}
+impl IcmRegionNum for Region2 {
+    const NUM: usize = 2;
+    const OFFSET: u32 = 0x10 * 2;
+}
+
+/// ICM Region 3
+pub enum Region3 {}
+impl Sealed for Region3 {}
+impl IcmRegionNum for Region3 {
+    const NUM: usize = 3;
+    const OFFSET: u32 = 0x10 * 3;
+}
+
+/// Functions required by [`IcmRegionDesc`]
+///
+/// Both Main List descriptors and Secondary List descriptors
+pub trait IcmRegionDescT {
+    /// Set the [`Raddr`] start of the region memory region
```

Comment:
I don't really like adding a `T` suffix to break name collisions between a type and a trait. I think it indicates that something isn't named as clearly as it could be.

In this case, it looks like you have `Main` descriptors and `Secondary` descriptors. Why not change `IcmRegionDesc` to `icm::MainRegionDesc`. I think that would match well with the `icm::SecondaryRegionDesc`.

Also, does the datasheet use Main & Secondary, not Primary & Secondary? If so, that's unfortunate. The pairing of Main & Secondary is a bit odd.

Reply:
Datasheet indeed uses Main and Secondary, example page 628 figure 26-2 and 26-3.

It became nicer with:

* `MainRegionDesc`
* `SecondaryRegionDesc`

which also frees up `RegionDesc` for the trait